### PR TITLE
StudyQuerySchema contextual role

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
The study schema currently enforces non-readability of datasets by failing DatasetDefinition.getTableInfo().  If we want to allow control of readability by other means (other than just the study security configuration), we need to allow for the TableInfo to be checked for permissions later (e.g.  UserSchema.getTable()).

A side effect of the current implementation is that many tables did not correctly implement canRead() as they simply would never be returned to the user if canRead()==false.

This change generalizes how we elevate permissions in the StudySchema to use contextual roles and also correctly implements canRead() for tables in the study schema.

#### Related Pull Requests
* 

#### Changes
* StudyQuerySchema can take a contextual role in its constructor (replaces mustCheckPermissions flag)
* Study tables consistently implement canRead()
* The study schema is now responsible for hiding tables w/o read permission.

### CODE REVIEW

**Comment by labkey-adam** [comment]
Build failing in `nircEHRModules` (newer repo) due to `canReadSchema()`: "attempting to assign weaker access privileges; was public"

File: api/src/org/labkey/api/data/DataRegion.java [hunk 1]
```diff
@@ -776,10 +777,11 @@ final public Results getResults(RenderContext ctx) throws SQLException, IOExcept
 
         try
         {
+            TableInfo tinfoMain = getTable();
```
Comment (labkey-adam) [inline api/src/org/labkey/api/data/DataRegion.java:780]: Why pull this up? Is it doing a required permissions check or something?

File: api/src/org/labkey/api/data/DataRegion.java [hunk 2]
```diff
@@ -2003,7 +2003,11 @@ private void renderUpdateForm(RenderContext ctx, Writer out) throws IOException
             }
             else
             {
-                Collection<Map<String, Object>> maps = new TableSelector(getTable(), selectKeyMap.values(), new PkFilter(getTable(), viewForm.getPkVals()), null).getMapCollection();
+                if (!hasPermission(ctx, ReadPermission.class))
+                    throw new UnauthorizedException();
+
+                TableInfo tinfoMain = getTable();
+                Collection<Map<String, Object>> maps = new TableSelector(tinfoMain, selectKeyMap.values(), new PkFilter(getTable(), viewForm.getPkVals()), null).getMapCollection();
```
Comment (labkey-adam) [inline api/src/org/labkey/api/data/DataRegion.java]:
```suggestion
                Collection<Map<String, Object>> maps = new TableSelector(tinfoMain, selectKeyMap.values(), new PkFilter(tinfoMain, viewForm.getPkVals()), null).getMapCollection();
```

**Comment by labkey-adam** [comment]
@labkey-matthewb Bunch of "use DatasetTableImpl.hasPermission()" exceptions in the BVT-EHR runs

File: study/src/org/labkey/study/query/StudyQuerySchema.java [hunk 3]
```diff
@@ -221,9 +248,11 @@ public static StudyQuerySchema createSchemaWithoutStudy(Container c, User u)
 
     protected Set<String> getSubSchemaNames()
     {
-        if (AppProps.getInstance().isExperimentalFeatureEnabled(EXPERIMENTAL_STUDY_SUBSCHEMAS))
-            return new LinkedHashSet<>(Arrays.asList("Datasets","Design","Specimens"));
-        return Collections.emptySet();
+        if (!AppProps.getInstance().isExperimentalFeatureEnabled(EXPERIMENTAL_STUDY_SUBSCHEMAS))
+            return Collections.emptySet();
+        if (null != ModuleLoader.getInstance().getModule("Specimen"))
```
Comment (labkey-adam) [inline study/src/org/labkey/study/query/StudyQuerySchema.java:251]: You can check `null != SpecimenService.get()` instead

**Review by labkey-adam — CHANGES_REQUESTED** [review]
Everything reviewed and seems very close. Please review comments and get EHR, etc. tests in reasonable shape and I'll approve.

File: api/src/org/labkey/api/data/DataRegion.java (hunk 1 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/data/DataRegion.java:780]: I think that was an intermediate change before I switched this.hasPermission()

Comment (labkey-matthewb) [reply api/src/org/labkey/api/data/DataRegion.java:780]: changed in 21.11_fb_readerrole

**Pull request closed without merging**